toolkit/data-sources/db-sync/src/governed_map/mod.rs

```diff
@@ -88,3 +92,229 @@ impl GovernedMapDataSourceImpl {
+
+			let changes = self.get_changes_in_range_to_cache(since_block_number, max_search_block, scripts).await?;
+
+			if let Ok(mut cache) = self.cache.lock() {
+				cache.add_changes(changes.clone());
+			}
+
+			Ok(deduplicate_changes(filter_changes_in_range(changes, since_block_number, up_to_block.block_no)))
+		}
+	}
+);
+
+impl GovernedMapDataSourceCachedImpl {
+	pub fn new_from_env(
+		pool: PgPool,
+		metrics_opt: Option<McFollowerMetrics>,
+	) -> std::result::Result<Self, &'static str> {
+		let security_parameter: u32 = std::env::var("CARDANO_SECURITY_PARAMETER")
+			.ok()
+			.and_then(|s| s.parse().ok())
+			.ok_or("Couldn't read env variable CARDANO_SECURITY_PARAMETER as u32")?;
+		Ok(Self {
+			pool,
+			metrics_opt,
+			security_parameter,
+			cache_size: 1000,
+			cache: Default::default(),
+		})
+	}
+
+	async fn get_changes_from_cache(
+		&self,
+		since_block: Option<BlockNumber>,
+		up_to_block: BlockNumber,
+	) -> Result<Option<Vec<(String, Option<ByteString>)>>> {
+		if let Ok(cache) = self.cache.lock() {
+			if let Some(changes) = cache.get_changes_in_range(since_block, up_to_block) {
+				return Ok(Some(deduplicate_changes(changes)));
+			}
+		};
+		return Ok(None);
+	}
+
+	async fn get_changes_in_range_to_cache(
+		&self,
+		since_block: Option<BlockNumber>,
+		up_to_block: BlockNumber,
+		scripts: MainChainScriptsV1,
+	) -> Result<Vec<(String, Option<ByteString>, BlockNumber, TxIndexInBlock)>> {
+		let latest_block_number = {
+			let cache = self.cache.lock().expect("Failed to lock cache");
+
+			cache.newest_block_number.or(since_block)
+		};
+		let changes = crate::db_model::get_changes(
+			&self.pool,
+			&scripts.validator_address.into(),
+			latest_block_number,
+			up_to_block,
+			Asset::new(scripts.asset_policy_id),
+		)
+		.await?;
+
+		let mut result = Vec::new();
+
+		for change in changes {
+			match GovernedMapDatum::try_from(change.datum.0) {
+				Ok(GovernedMapDatum { key, value }) => match change.action.as_str() {
+					"remove" => result.push((key, None, change.block_no, change.block_index)),
+					"upsert" => {
+						result.push((key, Some(value), change.block_no, change.block_index))
+					},
+					_ => warn!("Unknown action: {}", change.action),
+				},
+				Err(err) => warn!("Failed decoding map entry: {err}"),
+			}
+		}
+		Ok(result)
+	}
+}
+
+#[derive(Default)]
+pub(crate) struct Cache {
+	newest_block_number: Option<BlockNumber>,
+	changes: Vec<(String, Option<ByteString>, BlockNumber, TxIndexInBlock)>,
+	address: Option<MainchainAddress>,
+	policy_id: Option<PolicyId>,
+}
+
+fn deduplicate_changes(
+	mut changes: Vec<(String, Option<ByteString>, BlockNumber, TxIndexInBlock)>,
+) -> Vec<(String, Option<ByteString>)> {
+	let mut result = HashMap::new();
+
+	changes.sort_by_key(|(_, _, block_number, block_index)| (*block_number, *block_index));
+
+	for (key, value, _, _) in changes {
+		if result.contains_key(&key) {
```

Comment:
Is it necessary? Wouldn't `insert` overwrite anyway?

Reply:
I think there is one liner for creating this hash map:
`let m: HashMap<_, _> = changes.into_iter().map(|(key, value, _, _)|(key, value)).collect()`